auth_service/src/utility.rs

```diff
@@ -0,0 +1,140 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#![allow(dead_code)]
+use std::sync::Arc;
+
+use actix_web::{http, web, HttpRequest, HttpResponse};
+use crypto::Hasher;
+use shared::dto_models;
+use shared::sql::{Database, DatabaseResult};
+
+use crate::crypto;
+
+pub fn single_error(status: u16, detail: &str) -> HttpResponse {
+    let response =
+        dto_models::ErrorsResponse::default().with_error(dto_models::Error::default().status(status).detail(detail));
+
+    HttpResponse::build(http::StatusCode::from_u16(status).unwrap()).json(response)
+}
+
+
+pub fn resolve_database_entry<T>(result: DatabaseResult<T>, resource_name: &str) -> Result<T, HttpResponse> {
+    match result {
+        Ok(Some(entry)) => Ok(entry),
+        Ok(None) => Err(single_error(404, &format!("{} not found", resource_name))),
+        Err(error) => {
+            log::error!("Failed to get entry from SQL database due to {}", error);
+
+            Err(single_error(500, "Database lookup failed"))
+        }
+    }
+}
+
+
+pub async fn resolve_user(
+    req: &HttpRequest,
+    db: &web::Data<Arc<dyn Database>>,
+    hasher: &web::Data<Arc<dyn Hasher>>
+) -> Result<shared::dao_models::AuthUser, HttpResponse> {
+    let value = req
+        .headers()
+        .get(http::header::AUTHORIZATION)
+        .ok_or_else(|| {
+            let response = dto_models::ErrorsResponse::default().with_error(
+                dto_models::Error::default()
+                    .status(http::StatusCode::UNAUTHORIZED.as_u16())
+                    .detail("Missing authorization header")
+            );
+
+            HttpResponse::build(http::StatusCode::UNAUTHORIZED)
+                .insert_header((http::header::WWW_AUTHENTICATE, "Basic"))
+                .json(response)
+        })?
+        .to_str()
+        .map_err(|_| single_error(400, "Invalid authorization header"))?
+        .to_owned();
+
+    if value.len() < 7 {
+        return Err(single_error(400, "Invalid authorization header"));
+    }
+
+    let (token_type, token) = value.split_at(6);
+
+    if !"Basic ".eq_ignore_ascii_case(token_type) {
+        return Err(single_error(401, "Expected a Basic authorization token"));
+    }
+
+    let token = sodiumoxide::base64::decode(token, sodiumoxide::base64::Variant::Original)
```

Comment:
Lmao invalid auth header should be a 401 (also make helper functions for 401 and 500)